### PR TITLE
fix: remove listeners which doesn't clear correctly

### PR DESCRIPTION
**Pull request by Yuvalke**
### Description of the Changes

Didn't finished yet

### CheckLists

- [ ] changes have been done against master branch, and PR does not conflict
- [ ] new unit / functional tests have been added (whenever applicable)
- [ ] test are passing in local environment
- [ ] Travis tests are passing (or test results are not worse than on master branch :))
- [ ] Docs have been updated

### CODE REVIEW

File: src/plugin/base-plugin.js [hunk 1]
```diff
@@ -136,7 +138,9 @@ export default class BasePlugin implements IPlugin {
    * @virtual
    * @returns {void}
    */
-  reset(): void {}
+  reset(): void {
+    this.eventManager.removeAll();
```
Comment (OrenMe) [inline src/plugin/base-plugin.js:142]: it is up to plugin to decide this

File: src/engines/engine-decorator.js [hunk 2]
```diff
@@ -48,6 +50,15 @@ class EngineDecorator extends FakeEventTarget implements IEngineDecorator {
     return activeDecorator ? activeDecorator.dispatchEvent && activeDecorator.dispatchEvent(event) : super.dispatchEvent(event);
   }
 
+  _reset(): void {
+    this._eventManager.removeAll();
```
Comment (OrenMe) [inline src/engines/engine-decorator.js:54]: same here, do we event reset the proxy and re-register to the events? 
Also, the dependency is backward here, player creates this instance so need to call from player on destroy of proxy and not listen to events.

File: src/engines/html5/media-source/base-media-source-adapter.js [hunk 3]
```diff
@@ -186,11 +187,13 @@ export default class BaseMediaSourceAdapter extends FakeEventTarget implements I
    * @private
    */
   _handleLiveTimeUpdate(): void {
-    this._videoElement.addEventListener(Html5EventType.DURATION_CHANGE, () => {
-      if (this.isLive() && this._videoElement.paused) {
-        this._trigger(Html5EventType.TIME_UPDATE);
-      }
-    });
+    this._videoElement.addEventListener(Html5EventType.DURATION_CHANGE, this._onDurationChanged.bind(this));
```
Comment (RoyBregman) [inline src/engines/html5/media-source/base-media-source-adapter.js:190]: Why not use event manager also here?

File: src/engines/html5/media-source/adapters/fairplay-drm-handler.js [hunk 4]
```diff
@@ -114,7 +117,7 @@ class FairplayDrmHandler {
     let message = event.message;
     let request = new XMLHttpRequest();
     request.responseType = 'arraybuffer';
-    request.addEventListener('load', (e: Event) => this._licenseRequestLoaded(e), false);
+    request.addEventListener('load', (e: Event) => this._licenseRequestLoaded(e));
```
Comment (RoyBregman) [inline src/engines/html5/media-source/adapters/fairplay-drm-handler.js:120]: Why not use event manager also here?

Comment (RoyBregman) [inline src/engines/html5/media-source/adapters/fairplay-drm-handler.js:120]: Why not use event manager also here?

File: src/engines/html5/media-source/base-media-source-adapter.js (hunk 3 above)
Comment (Yuvalke) [reply src/engines/html5/media-source/base-media-source-adapter.js:190]: If media source adapter will do remove all listener this listener will be deleted and nobody will re-register it. 
I think we want to make sure it'll not happened

File: src/engines/html5/media-source/adapters/fairplay-drm-handler.js (hunk 4 above)
Comment (Yuvalke) [reply src/engines/html5/media-source/adapters/fairplay-drm-handler.js:120]: Could be many requests and after it's finish the request object is not relevant anymore, it's a local var.
It'll fill the event manager with non relevant handlers

Comment (RoyBregman) [reply src/engines/html5/media-source/adapters/fairplay-drm-handler.js:120]: I don't understand, like the way it is now the listener will be never removed...

Comment (Yuvalke) [reply src/engines/html5/media-source/adapters/fairplay-drm-handler.js:120]: Right, it should be removed.

File: src/engines/engine-decorator.js [hunk 5]
```diff
@@ -22,6 +22,7 @@ class EngineDecorator extends FakeEventTarget implements IEngineDecorator {
   constructor(engine: IEngine, pluginWithDecorators: Array<IEngineDecoratorProvider>) {
     super();
     this._eventManager = new EventManager();
+    this._eventManager.listen(engine, EventType.PLAYER_DESTROY, () => this._destroy());
```
Comment (OrenMe) [inline src/engines/engine-decorator.js:25]: @Yuvalke why listen to the underlying engine when this is explicitly called by the player?
Add this as a method and call the underlying engine explicitly.

File: src/engines/html5/media-source/base-media-source-adapter.js (hunk 3 above)
Comment (OrenMe) [reply src/engines/html5/media-source/base-media-source-adapter.js:190]: this won't remove the listener, you need to keep reference to the bounded method and use that in the add and remove

**Pull request closed without merging**